File: Backend/src/main/java/com/woovi/crudbank/shared/util/AccountFieldNormalizer.java
package com.woovi.crudbank.shared.util;

import com.woovi.crudbank.shared.error.DomainException;

import java.util.regex.Pattern;

public final class AccountFieldNormalizer {

    private static final Pattern ACCOUNT_NUMBER_PATTERN = Pattern.compile("^\\d{5,12}(-\\d{1})?$");

    private AccountFieldNormalizer() {
    }

    public static String normalizeOwnerName(String ownerName) {
        if (ownerName == null || ownerName.trim().isEmpty()) {
            throw DomainException.validationField("ownerName", "Owner name is required");
        }

        String normalized = ownerName.trim().replaceAll("\\s+", " ");
        if (normalized.length() < 3 || normalized.length() > 120) {
            throw DomainException.validationField("ownerName", "Owner name must be between 3 and 120 characters");
        }

        return normalized;
    }

    public static String normalizeDocument(String document) {
        return DocumentValidator.normalizeAndValidateBrazilDocument(document);
    }

    public static String normalizeBranch(String branch) {
        if (branch == null || branch.trim().isEmpty()) {
            throw DomainException.validationField("branch", "Branch is required");
        }

        String normalized = branch.replaceAll("\\D", "");
        if (normalized.length() != 4) {
            throw DomainException.validationField("branch", "Branch must contain exactly 4 digits");
        }

        return normalized;
    }

    public static String normalizeAccountNumber(String number) {
        if (number == null || number.trim().isEmpty()) {
            throw DomainException.validationField("number", "Account number is required");
        }

        String normalized = number.trim().replaceAll("\\s+", "");
        if (!ACCOUNT_NUMBER_PATTERN.matcher(normalized).matches()) {
            throw DomainException.validationField("number", "Account number must match 12345 or 12345-6 pattern");
        }

        return normalized;
    }
}

File: backend-js/src/shared/validators/accountFieldNormalizer.ts
import { DomainError } from '../errors/DomainError';
import { normalizeAndValidateBrazilDocument } from './documentValidator';

const ACCOUNT_NUMBER_PATTERN = /^\d{5,12}(-\d{1})?$/;

export function normalizeOwnerName(ownerName: string): string {
  if (!ownerName || ownerName.trim().length === 0) {
    throw DomainError.validationField('ownerName', 'Owner name is required');
  }

  const normalized = ownerName.trim().replace(/\s+/g, ' ');
  if (normalized.length < 3 || normalized.length > 120) {
    throw DomainError.validationField('ownerName', 'Owner name must be between 3 and 120 characters');
  }

  return normalized;
}

export function normalizeDocument(document: string): string {
  return normalizeAndValidateBrazilDocument(document);
}

export function normalizeBranch(branch: string): string {
  if (!branch || branch.trim().length === 0) {
    throw DomainError.validationField('branch', 'Branch is required');
  }

  const normalized = branch.replace(/\D/g, '');
  if (normalized.length !== 4) {
    throw DomainError.validationField('branch', 'Branch must contain exactly 4 digits');
  }

  return normalized;
}

export function normalizeAccountNumber(number: string): string {
  if (!number || number.trim().length === 0) {
    throw DomainError.validationField('number', 'Account number is required');
  }

  const normalized = number.trim().replace(/\s+/g, '');
  if (!ACCOUNT_NUMBER_PATTERN.test(normalized)) {
    throw DomainError.validationField('number', 'Account number must match 12345 or 12345-6 pattern');
  }

  return normalized;
}
